k2/csrc/fsa_utils.cu

```diff
@@ -99,75 +99,237 @@ static void SplitStringToVector(const std::string &in, const char *delim,
-      // we first have to collect all the final states, and then work out the
-      // super final state.
-      K2_CHECK(openfst) << "Invalid line: " << line
-                        << "\nFinal state with weight detected in K2 format";
+      float score = 0.0f;
+      if (num_fields == 2u)
+        score = StringToFloat(splits[1]);
       original_final_states.push_back(StringToInt(splits[0]));
-      original_final_weights.push_back(StringToFloat(splits[1]));
+      original_final_weights.push_back(score);
       max_state = std::max(max_state, original_final_states.back());
-    } else if (num_fields == 1u) {
-      //   0
-      // final_state
-      (void)StringToInt(splits[0]);  // this is a final state
-      break;                         // finish reading
     } else {
       K2_LOG(FATAL) << "Invalid line: " << line
-                    << "\nIt expects a line with 1, 2 or 4 fields";
+                    << "\nOpenFST acceptor expects a line with 1 (final_state),"
+                    << " 2 (final_state score), 3 (src_state dest_state label) "
+                    << "or 4 (src_state dest_state label score) fields.";
     }
-  } while (std::getline(is, line));
+  }
 
-  // Post processing on final states. When openfst is true, we may have multiple
-  // final states with weights associated with them. We will have to add a super
-  // final state, and convert that into the K2 format (final state with no
+  // Post processing on final states. We may have multiple final states with
+  // weights associated with them for OpenFST style FSTs. We will have to add a
+  // super final state, and convert that into the K2 format (final state with no
   // weight).
   if (original_final_states.size() > 0) {
-    K2_CHECK_EQ(openfst, true);
     K2_CHECK_EQ(original_final_states.size(), original_final_weights.size());
     int32_t super_final_state = max_state + 1;
+    state_to_arcs.resize(super_final_state);
     for (std::size_t i = 0; i != original_final_states.size(); ++i) {
-      arcs.emplace_back(original_final_states[i], super_final_state,
-                        -1,  // kFinalSymbol
-                        scale * original_final_weights[i]);
+      state_to_arcs[original_final_states[i]].emplace_back(
+          original_final_states[i], super_final_state,
+          -1,  // kFinalSymbol
+          -1.0f * original_final_weights[i]);
+      ++num_arcs;
     }
   }
-  if (openfst) {
-    // Sort arcs so that source states are in non-decreasing order.
-    std::sort(arcs.begin(), arcs.end());
+
+  // Move arcs from "state_to_arcs" to "arcs".
+  std::size_t arc_index = 0;
+  arcs.resize(num_arcs);
+  for (std::size_t s = 0; s < state_to_arcs.size(); ++s) {
+    for (std::size_t a = 0; a < state_to_arcs[s].size(); ++a) {
+      K2_CHECK_GT(num_arcs, arc_index);
+      arcs[arc_index] = state_to_arcs[s][a];
+      ++arc_index;
+    }
   }
+  K2_CHECK_EQ(num_arcs, static_cast<int32_t>(arc_index));
+  state_to_arcs.resize(0);
```

Comment:
`resize(0)` will not free the underlying memory.
This statement can be deleted.